### DISCUSSION TITLE
matlab at matching as fast as MAD

### DISCUSSION THREAD

**Discussion by simoneliuzzo** [atcollab/at#445] [Matlab AT]
The AT matching is really slow compared to MAD matching. 
Is there anything that could be done to make it faster (100 time faster)? 
In MAD8 the EBS optics matching takes an imperceptible time. In AT several minutes.

**Comment by swhite2401** [comment]
@liuzzo, very good point and long standing issue!
I am not an expert of MAD8 but there are few things to know about MADX, the default calculation assumes:
-no coupling
-no radiation
-no chromatic functions
-with `BETA0=XXX` a line is considered, no closed orbit search
-for periodic solutions, I am not sure, but it is well possible that if the closed orbit is zero the search does not take place a every loop.

In pyAT, I could approach MADX speed in computing the optics using a transfer line or define `orbit=numpy.zeros(6)` and the uncoupled lattice calculation (`at.linopt2`). 
For the closed orbit search I am not sure how the codes compare (but for zero closed orbit it should be very fast in principle...)
Then the least square fitting routine is a numpy function and there is nothing we can do about it...

Do you have a test example we could use for benchmarking?

**Comment by lfarv** [comment]
Hello,
Accelerating the matching involves two elements:
- the minimiser: `fminsearch` or `lsqnonlin` in Matlab, `'trf'` of `'lm'` in python (scipy). Though I have no reason to believe that Matlab is particularly bad in that, it would be interesting to compare the efficiency of these 4 methods on an identical evaluation function (analytical multi-variable function for instance),
- the evaluation function. For closed orbit, linear optics or envelope computation, I did not see any significant difference in speed between Matlab and python. However, building the evaluation function is quite different. For Matlab, @simoneliuzzo is the best person to look at that and look for redundant or useless computations.

@swhite2401: in my experience, closed orbit computation is very fast, compared to optics. But you are right, freezing the orbit a zero if you know that there is no closed orbit is a good thing to do.
